cluster/assemblestate/assertion.go

```diff
@@ -0,0 +1,138 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2025 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package assemblestate
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/snapcore/snapd/asserts"
+)
+
+// AssertionDevices converts the data returned by an assembly session into the
+// data structure used by the "devices" block of a cluster assertion.
+func AssertionDevices(ids []Identity, routes Routes) ([]any, error) {
+	addresses, err := addressesFromRoutes(routes)
+	if err != nil {
+		return nil, err
+	}
+
+	devices := make([]any, 0, len(ids))
+	for i, identity := range ids {
+		serial, err := serialFromBundle(identity.SerialBundle)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse serial bundle for device %q: %w", identity.RDT, err)
+		}
+
+		addrs := append([]string(nil), addresses[identity.RDT]...)
+		if len(addrs) == 0 {
+			return nil, fmt.Errorf("no addresses available for device %q", identity.RDT)
+		}
+
+		header := make([]any, 0, len(addrs))
+		for _, addr := range addrs {
+			header = append(header, addr)
+		}
+
+		devices = append(devices, map[string]any{
+			"id":        strconv.Itoa(i + 1),
```

Comment:
would it make sense to sort the Identitys in some way to give more predictable id assignments here?